src/System.Runtime.Caching/src/System/Runtime/Caching/CacheMemoryMonitor.cs

```diff
@@ -0,0 +1,297 @@
+                    Interlocked.Exchange(ref s_effectiveProcessMemoryLimit, memoryLimit);
+                }
+                return memoryLimit;
+            }
+        }
+
+        protected override int GetCurrentPressure()
+        {
+            // Call GetUpdatedTotalCacheSize to update the total
+            // cache size, if there has been a recent Gen 2 Collection.
+            // This update must happen, otherwise the CacheManager won't 
+            // know the total cache size.
+            int gen2Count = GC.CollectionCount(2);
+            SRefMultiple sref = _sizedRefMultiple;
+            if (gen2Count != _gen2Count && sref != null)
+            {
+                // update _gen2Count
+                _gen2Count = gen2Count;
+
+                // the SizedRef is only updated after a Gen2 Collection
+
+                // increment the index (it's either 1 or 0)
+                Dbg.Assert(SAMPLE_COUNT == 2);
+                _idx = _idx ^ 1;
+                // remember the sample time
+                _cacheSizeSampleTimes[_idx] = DateTime.UtcNow;
+                // remember the sample value
+                _cacheSizeSamples[_idx] = sref.ApproximateSize;
+#if DBG
+                Dbg.Trace("MemoryCacheStats", "SizedRef.ApproximateSize=" + _cacheSizeSamples[_idx]);
+#endif
+                IMemoryCacheManager memoryCacheManager = s_memoryCacheManager;
+                if (memoryCacheManager != null)
+                {
+                    memoryCacheManager.UpdateCacheSize(_cacheSizeSamples[_idx], _memoryCache);
+                }
+            }
+
+            // if there's no memory limit, then there's nothing more to do
+            if (_memoryLimit <= 0)
+            {
+                return 0;
+            }
+
+            long cacheSize = _cacheSizeSamples[_idx];
+
+            // use _memoryLimit as an upper bound so that pressure is a percentage (between 0 and 100, inclusive).
+            if (cacheSize > _memoryLimit)
+            {
+                cacheSize = _memoryLimit;
+            }
+
+            // PerfCounter: Cache Percentage Process Memory Limit Used
+            //    = memory used by this process / process memory limit at pressureHigh
+            // Set private bytes used in kilobytes because the counter is a DWORD
+
+            // PerfCounters.SetCounter(AppPerfCounter.CACHE_PERCENT_PROC_MEM_LIMIT_USED, (int)(cacheSize >> KILOBYTE_SHIFT));
+
+            int result = (int)(cacheSize * 100 / _memoryLimit);
+            return result;
+        }
+
+        internal override int GetPercentToTrim(DateTime lastTrimTime, int lastTrimPercent)
+        {
+            int percent = 0;
+            if (IsAboveHighPressure())
+            {
+                long cacheSize = _cacheSizeSamples[_idx];
+                if (cacheSize > _memoryLimit)
+                {
+                    percent = Math.Min(100, (int)((cacheSize - _memoryLimit) * 100L / cacheSize));
+                }
+
+#if PERF
```

Comment:
You might want to change all OutputDebugString to Debug.WriteLine, which will do the same thing by default, but is more configurable,